watertrading/code/BuildModel.py

```diff
@@ -0,0 +1,1361 @@
+                                for ind,row in df_vM_Truck.iterrows()]
+        df_vM_Truck["Departure Date"] = df_vM_Truck["Departure Date"].dt.strftime("%m/%d/%Y")
+        df_vM_Truck["Arrival Date"] = df_vM_Truck["Arrival Date"].dt.strftime("%m/%d/%Y")
+        df_vM_Truck["From Longitude"] = [model.p_ProducerWellpadLon[pad] for pad in df_vM_Truck.loc[:,"From wellpad"]]
+        df_vM_Truck["From Latitude"] = [model.p_ProducerWellpadLat[pad] for pad in df_vM_Truck.loc[:,"From wellpad"]]
+        df_vM_Truck["To Longitude"] = [model.p_ConsumerWellpadLon[pad] for pad in df_vM_Truck.loc[:,"To wellpad"]]
+        df_vM_Truck["To Latitude"] = [model.p_ConsumerWellpadLat[pad] for pad in df_vM_Truck.loc[:,"To wellpad"]]
+        df_vM_Truck["Operator"] = [model.p_MidstreamOperator[index] for index in df_vM_Truck["Midstream Index"]]
+
+    df_v_Supply = pd.DataFrame({"Supplier Index": key[0],
+                                "Supplier Wellpad": model.p_ProducerPad[key[0]],
+                                "Date Index": key[1],
+                                "Date": model.d_T[key[1]],
+                                "Rate": value(model.v_Supply[key])}
+                                for key in model.v_Supply
+                                if value(model.v_Supply[key])>0)
+    if not df_v_Supply.empty:
+        df_v_Supply["Date"]= df_v_Supply["Date"].dt.strftime("%m/%d/%Y")
+        df_v_Supply["Longitude"] = [model.p_ProducerWellpadLon[pad] for pad in df_v_Supply.loc[:,"Supplier Wellpad"]]
+        df_v_Supply["Latitude"] = [model.p_ProducerWellpadLat[pad] for pad in df_v_Supply.loc[:,"Supplier Wellpad"]]
+        df_v_Supply["Nodal Price"] = add_dataframe_prices(df_v_Supply, model.p_ProducerNodalPrice, "Supplier")
+        df_v_Supply["Operator"] = [model.p_ProducerOperator[index] for index in df_v_Supply["Supplier Index"]]
+
+    df_v_Demand = pd.DataFrame({"Consumer Index": key[0],
+                                "Consumer Wellpad": model.p_ConsumerPad[key[0]],
+                                "Date Index": key[1],
+                                "Date": model.d_T[key[1]],
+                                "Rate": value(model.v_Demand[key])}
+                                for key in model.v_Demand
+                                if value(model.v_Demand[key])>0)
+    if not df_v_Demand.empty:
+        df_v_Demand['Date']= df_v_Demand['Date'].dt.strftime("%m/%d/%Y")
+        df_v_Demand["Longitude"] = [model.p_ConsumerWellpadLon[pad] for pad in df_v_Demand.loc[:,"Consumer Wellpad"]]
+        df_v_Demand["Latitude"] = [model.p_ConsumerWellpadLat[pad] for pad in df_v_Demand.loc[:,"Consumer Wellpad"]]
+        df_v_Demand["Nodal Price"] = add_dataframe_prices(df_v_Demand, model.p_ConsumerNodalPrice, "Consumer")
+        df_v_Demand["Operator"] = [model.p_ConsumerOperator[index] for index in df_v_Demand["Consumer Index"]]
+
+    if df_v_Supply.empty & df_v_Demand.empty:
```

Comment:
```suggestion
    if df_v_Supply.empty and df_v_Demand.empty:
```
I think you want to use `and` here instead of `&`. In Python `and` is a logical operator and `&` is a bitwise operator (https://www.geeksforgeeks.org/difference-between-and-and-in-python/). I think the former is what you want here. `&` can also be used for the set intersection operation, but I'm pretty sure that is not what you are doing here.

Reply:
addressed